### PR TITLE
Add tutorial of modality crossing from text to image

### PR DESCRIPTION
**Pull request by tqtg**
### Description

This tutorial shows how to crossing auxiliary data between text modality to image modality. VBPR model is used as an example.

### Checklist:

- [ ] I have added tests.
- [ ] I have updated the documentation accordingly.
- [ ] I have updated `README.md` (if you are adding a new model).
- [ ] I have updated `examples/README.md` (if you are adding a new example).

### CODE REVIEW

File: tutorials/vbpr_text.ipynb [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "*Copyright (c) Cornac Authors. All rights reserved.*\n",
+    "\n",
+    "*Licensed under the Apache 2.0 License.*\n",
+    "\n",
+    "# Visual Bayesian Personalizer Ranking (VBPR) with Text Data"
```
Comment (saghiles) [inline tutorials/vbpr_text.ipynb]: Move the acronym VBPR to the main text

File: tutorials/vbpr_text.ipynb [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+    "</table>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Overview\n",
+    "In this tutorial, we will employ a recommendation algorithm designed to leverage item images to work with item text content.\n",
+    "The model of interest is [VBPR](https://arxiv.org/abs/1510.01784) that makes use of visual features extracted from item images using pre-trained CNN deep networks.\n",
```
Comment (saghiles) [inline tutorials/vbpr_text.ipynb]: Why not link directly to the pdf, is it related to the loading time?

Comment (saghiles) [inline tutorials/vbpr_text.ipynb]: "The model of interest is VBPR, which makes use ..." or "... VBPR making use ...". "That" is not suitable here cause there is only one VBPR.

File: tutorials/vbpr_text.ipynb [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+    "\n",
+    "What if our data doesn't come with item images but text, for example, we are working with [MovieLens dataset](https://grouplens.org/datasets/movielens/) that contains text plots of the movies. At the same time, we also want to work with VBPR model which is not originally designed for text data. To address that challenge, Cornac provides mechanisms to cross between modalities, in which here we focus on crossing from text to image to utilize image models for text-available datasets.\n",
```
Comment (saghiles) [inline tutorials/vbpr_text.ipynb]: "What if our data doesn't come with item images but text`?`" For example, this is the case of the MovieLens dataset containing text movie plots.

File: tutorials/vbpr_text.ipynb (hunk 1 above)
Comment (tqtg) [reply tutorials/vbpr_text.ipynb]: updated

File: tutorials/vbpr_text.ipynb (hunk 2 above)
Comment (tqtg) [reply tutorials/vbpr_text.ipynb]: updated

Comment (tqtg) [reply tutorials/vbpr_text.ipynb]: updated

File: tutorials/vbpr_text.ipynb (hunk 3 above)
Comment (tqtg) [reply tutorials/vbpr_text.ipynb]: It's a phrase to state possibility, not a question.

File: tutorials/vbpr_text.ipynb [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+    "  </td>\n",
+    "</table>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Overview\n",
+    "In this tutorial, we will employ a recommendation algorithm designed to leverage item images to work with item text content.\n",
```
Comment (saghiles) [inline tutorials/vbpr_text.ipynb]: It seems like the first two paragraphs are closely related, and it could be better if they can be summarized into one. In particular, the question regarding the availability of image data can come earlier. I suggest something like this: "We are interested in the following situation: we would like to use the Visual Bayesian Personalized Ranking (VBPR) model, which leverages pre-trained item visual features to improve recommendation. However, our data of interest, namely MovieLens, does not come with visual information, but instead it contains text movie plots.  In this tutorial, we will see how Conac's *modality* infrastructures allow us to easily use VBPR (originally designed for visual information) to leverage item text content." If you choose to adopt this suggestion, my comments below on the first two paragraphs become useless, so please feel free to ignore them.

File: tutorials/vbpr_text.ipynb [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+    "from cornac.data import TextModality, ImageModality\n",
+    "from cornac.data.text import BaseTokenizer"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Prepare data\n",
+    "Here we use MovieLens 100K dataset as already provided inside Cornac library. We load movie plots with their corresponding ids, then load the rating data filtered the movies that don't come with plots."
```
Comment (saghiles) [inline tutorials/vbpr_text.ipynb]: Here we use `the` MovieLens 100K dataset, which is already accessible from Cornac. Hence, we can simply load the movie plots and user-movie rating data as follows:

File: tutorials/vbpr_text.ipynb [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# load movie plots from built-in movielens datasets\n",
```
Comment (saghiles) [inline tutorials/vbpr_text.ipynb]: Remove this comment

File: tutorials/vbpr_text.ipynb [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+    "plots, movie_ids = movielens.load_plot()\n",
+    "\n",
+    "# load rating data of MovieLens 100K\n",
```
Comment (saghiles) [inline tutorials/vbpr_text.ipynb]: Remove this comment

File: tutorials/vbpr_text.ipynb (hunk 4 above)
Comment (tqtg) [reply tutorials/vbpr_text.ipynb]: updated

File: tutorials/vbpr_text.ipynb [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+    "</table>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Overview\n",
+    "\n",
+    "We would like to use [Visual Bayesian Personalizer Ranking (VBPR)](https://arxiv.org/pdf/1510.01784.pdf), it makes use of pre-trained visual features extracted from CNN. However, our data of interest [MovieLens dataset](https://grouplens.org/datasets/movielens/) does not come with visual information, but instead it contains text movie plots. In this tutorial, we will employ Conac's modality infrastructures to easily utilize VBPR to leverage item text content.\n",
```
Comment (saghiles) [inline tutorials/vbpr_text.ipynb]: `which` makes use

File: tutorials/vbpr_text.ipynb [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+    "ml_100k = movielens.load_100k(reader=Reader(item_set=movie_ids)) "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Cross modality\n",
+    "\n",
+    "To get representations from text data, we build a `TextModality` using text corpus with coressponding ids as the input. We also need to supply a `Tokenizer` for text splitting, in this case the tokens in text plots are seperated by `\\tab` character. We limit the maximum size of vocabulary to 5000, which also means the size of representations (e.g., bag-of-words vectors) can not go beyond that limit."
```
Comment (saghiles) [inline tutorials/vbpr_text.ipynb]: To reduce repetitions in this paragraph, something like this may be considered: ''To get vector representations from raw texts, we rely on the `TextModality` class. In addition to our corpus and `corresponding` ids, we need to supply a `Tokenizer` for text splitting. We further set the maximum size of the vocabulary to 5000, which also means that the dimension of our vector space `cannot` go higher." Note that, have also highlighted some typos from the original text.

File: tutorials/vbpr_text.ipynb [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+    "item_text_modality = TextModality(corpus=plots, ids=movie_ids, \n",
+    "                                  tokenizer=BaseTokenizer(sep='\\t', stop_words='english'),\n",
+    "                                  max_vocab=5000, max_doc_freq=0.5).build()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "After building the `TextModality`, we can access text representations through provided APIs and attributes of the object. In this case, we take the word-count matrix as item representations then use it as input to construct an `ImageModality`. In other words, item visual representations will be substituted by bag-of-words representations which are used by `VBPR` model. "
```
Comment (saghiles) [inline tutorials/vbpr_text.ipynb]: This part could be made more concise as follows: "The next step is to instantiate an `ImageModality` (which VBPR relies on) using our text representations, i.e., the word-count matrix in this case.  In other words, we will substitute text representations for visual features."

File: tutorials/vbpr_text.ipynb [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+   "source": [
+    "features = item_text_modality.count_matrix.A\n",
+    "item_image_modality = ImageModality(features=features, ids=movie_ids)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In Cornac, image recommendation models always work with `ImageModality`, similarly for other modalities. That design makes it consistent with models' original assumptions, thus it's easy to integrate models into the library without any confusion of which modality to use.\n",
```
Comment (saghiles) [inline tutorials/vbpr_text.ipynb]: This text is a little bit hard to follow. How about something like this: "In Cornac, every model relies on the modality for which it was originally designed. For instance, `visual` recommendation algorithms always work with ImageModality. This ensures consistency with models' original assumptions, and helps us avoid confusions regarding which modality to use when integrating a new recommender model to Cornac."

File: tutorials/vbpr_text.ipynb [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+    "\n",
+    "## Experiment\n",
+    "\n",
+    "We employ `RatioSplit` evaluation method to split the rating data. The `item_image_modality` is also supplied here for later usage of model training."
```
Comment (saghiles) [inline tutorials/vbpr_text.ipynb]: We employ `the` RatioSplit evaluation method to split the rating data. The item_image_modality is also supplied here for later usage `by our model`. 

*Suggestion:* Another important point that could be mentioned here is, the fact that one does not have to worry about aligning the set of training items with the corresponding text features. This is probably the main advantage of providing the modality as input to the ratio split.

File: tutorials/vbpr_text.ipynb [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+    "                         item_image=item_image_modality,\n",
+    "                         exclude_unknowns=True, \n",
+    "                         verbose=True, seed=123)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "VBPR model is supported inside Cornac. Note that the implementation of VBPR uses [PyTorch](https://pytorch.org/), thus we need to install that dependency in order to run the model (done at the beginning). List of implemented models with their additional dependencies is carefully documented [here](https://github.com/PreferredAI/cornac#models).\n",
```
Comment (saghiles) [inline tutorials/vbpr_text.ipynb]: It may be necessary to recall the dependency of VBPR on PyTorch. This is done in the Setup section, and there is a need for more details on this aspect, it may be better to add them in the latter section (Setup).

This paragraph can then be reduced to something like: " We are now ready to fit and evaluate VBPR. We further include the BPR model as a baseline to assess the effectiveness of the text auxiliary data (movie plots). Both models are already implemented in Cornac thereby making such an evaluation straightforward".

File: tutorials/vbpr_text.ipynb [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+   "metadata": {},
+   "source": [
+    "Results after running the experiment:\n",
+    "\n",
+    "<pre>\n",
+    "     |    AUC | Recall@50 | Train (s) | Test (s)\n",
+    "---- + ------ + --------- + --------- + --------\n",
+    "BPR  | 0.7355 |    0.2297 |    1.3699 |  12.9997\n",
+    "VBPR | 0.7669 |    0.2499 |   61.1777 |  17.2591\n",
+    "</pre>"
```
Comment (saghiles) [inline tutorials/vbpr_text.ipynb]: *Suggestion:* say some words about the results and how they support cross-modality utilization of existing models.

File: tutorials/vbpr_text.ipynb (hunk 8 above)
Comment (tqtg) [reply tutorials/vbpr_text.ipynb]: noted

File: tutorials/vbpr_text.ipynb (hunk 10 above)
Comment (tqtg) [reply tutorials/vbpr_text.ipynb]: noted

File: tutorials/vbpr_text.ipynb (hunk 12 above)
Comment (tqtg) [reply tutorials/vbpr_text.ipynb]: noted

File: tutorials/vbpr_text.ipynb (hunk 13 above)
Comment (tqtg) [reply tutorials/vbpr_text.ipynb]: noted

File: tutorials/vbpr_text.ipynb (hunk 14 above)
Comment (tqtg) [reply tutorials/vbpr_text.ipynb]: noted

File: tutorials/vbpr_text.ipynb (hunk 9 above)
Comment (tqtg) [reply tutorials/vbpr_text.ipynb]: noted

File: tutorials/vbpr_text.ipynb (hunk 11 above)
Comment (tqtg) [reply tutorials/vbpr_text.ipynb]: noted

File: tutorials/vbpr_text.ipynb [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+    "                         item_image=item_image_modality,\n",
+    "                         exclude_unknowns=True, \n",
+    "                         verbose=True, seed=123)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We are now reading to evaluate performance of VBPR. We also include [BPR](https://arxiv.org/ftp/arxiv/papers/1205/1205.2618.pdf) model as a baseline to examine the effectiveness of the text auxiliary data."
```
Comment (saghiles) [inline tutorials/vbpr_text.ipynb]: We are now `ready` to evaluate performance of VBPR. We also include `the` BPR model

Comment (tqtg) [reply tutorials/vbpr_text.ipynb]: noted. thanks

File: tutorials/vbpr_text.ipynb [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+    "We are now ready to evaluate performance of VBPR. The [BPR](https://arxiv.org/ftp/arxiv/papers/1205/1205.2618.pdf) model is also included as a baseline to examine the effectiveness of the text auxiliary data."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "vbpr = cornac.models.VBPR(k=10, k2=10, n_epochs=20, batch_size=20, learning_rate=0.001,\n",
```
Comment (saghiles) [inline tutorials/vbpr_text.ipynb]: The hyperparameters (k, and n_epochs/max_iter) for VBPR and BPR are different. Can we choose these parameters in such away that VBPR collapses to BPR if remove the visual features? At least have the same `k`, it is fine to have different  `n_epochs/max_iter` since we are dealing with different objective functions.

Comment (saghiles) [reply tutorials/vbpr_text.ipynb]: Or may be the objective here is to make `k_vbpr+k2_vbpr = k_bpr` ?

Comment (tqtg) [reply tutorials/vbpr_text.ipynb]: Yes, it is. Anyway, I updated the tutorial.
Do you have any other comments?

**Comment by saghiles** [comment]
@tqtg no more comment on this pull request. Feel free to merge it when you are done
